File: Cisco/2BinaryTreeCameras.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def minCameraCover(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: int
        """
        self.num_of_cameras = 0

        def dfs(node):
            if not node:
                return False, True

            camera, monitored = False, False
            cleft, mleft = dfs(node.left)
            cright, mright = dfs(node.right)

            if cleft or cright:
                monitored = True

            if not mleft or not mright:
                self.num_of_cameras += 1
                camera = True
                monitored = True

            return camera, monitored

        _, monitored = dfs(root)
        return self.num_of_cameras if monitored else self.num_of_cameras + 1

File: Cisco/LinkedInList/C++/2BinaryTreeCameras.cpp
#include <tuple>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
    int numCameras;

public:
    int minCameraCover(TreeNode *root)
    {
        bool camera, monitored;
        tie(camera, monitored) = dfs(root);
        return monitored ? numCameras : numCameras + 1;
    }

    tuple<bool, bool> dfs(TreeNode *root)
    {
        if (!root)
            return {false, true};

        bool cameraLeft, monitoredLeft;
        tie(cameraLeft, monitoredLeft) = dfs(root->left);

        bool cameraRight, monitoredRight;
        tie(cameraRight, monitoredRight) = dfs(root->right);

        bool camera = !monitoredLeft || !monitoredRight,
             monitored = camera || cameraLeft || cameraRight;

        if (camera)
            numCameras++;

        return {camera, monitored};
    }
};